T2

maximKalinov opened the pull request:
Create domain entities

maximKalinov committed:
T1: Create application projects

maximKalinov committed:
T2: Create domain entities

maximKalinov committed:
T2: Configure entities

maximKalinov committed:
T2: Create DbContect

pipalalala commented on NugetAnalyzer.DAL/Context/NugetAnalyzerDbContext.cs:
```diff
@@ -0,0 +1,31 @@
+    public class NugetAnalyzerDbContext : DbContext
+    {
+        public NugetAnalyzerDbContext(DbContextOptions<NugetAnalyzerDbContext> options)
+            :base(options)
```
Please, add space before "base" keyword

pipalalala commented on NugetAnalyzer.DAL/EntityConfigurations/ReferencePackageConfiguration.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public void Configure(EntityTypeBuilder<ReferencePackage> builder)
+        {
+            builder.HasKey(k => k.PackageId);
```
Please, change "k" to "p"

pipalalala commented on NugetAnalyzer.DAL/EntityConfigurations/PackageConfiguration.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public void Configure(EntityTypeBuilder<Package> builder)
+        {
+            builder.HasKey(k => k.PackageId);
```
Please, change "k" to "p" everywhere

maximKalinov committed:
T2: Merge branch 'master' into domainEntitiesCreation/T2

# Conflicts:
#	NugetAnalyzer.DAL/NugetAnalyzer.DAL.csproj
#	NugetAnalyzer.sln
#	NugetAnalyzer/NugetAnalyzer.csproj
#	NugetAnalyzer/Program.cs
#	NugetAnalyzer/Startup.cs

maximKalinov committed:
T2: Fix references local system

maximKalinov committed:
T2: Comments resolving

HenadziBaranau commented on NugetAnalyzer.DAL/EntityConfigurations/ProjectConfiguration.cs:
```diff
@@ -0,0 +1,17 @@
+        {
+            builder.HasKey(p => p.ProjectId);
+
+            builder.Property(p => p.Name).IsRequired().HasMaxLength(4096);
```
Please allign similar to code below. IMHO, it's easier to read:
```
builder
      .Property(p => p.Name)
      .IsRequired()
      .HasMaxLength(4096);
```

maximKalinov replied:
Done

HenadziBaranau commented on NugetAnalyzer.DAL/Context/NugetAnalyzerDbContext.cs:
```diff
@@ -0,0 +1,31 @@
+        public DbSet<Project> Projects { get; set; }
+        public DbSet<Repository> Repositories { get; set; }
+        public DbSet<Solution> Solutions { get; set; }
+        public DbSet<User> Users { get; set; }
```
It's enough just to leave only `DbSet<User>`. All other collections are "reachable" from User base on navigation properties/foreign keys.

maximKalinov replied:
Done

HenadziBaranau commented on NugetAnalyzer.DAL/NugetAnalyzer.DAL.csproj:
```diff
@@ -4,6 +4,12 @@
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="2.2.4" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="2.2.4" />
```
Why do we need this package?

maximKalinov replied:
Remove it

HenadziBaranau commented on NugetAnalyzer.Domain/Package.cs:
```diff
@@ -0,0 +1,11 @@
+        public Project Project { get; set; }
+
+        public int? ReferencePackageId { get; set; }
+        public ReferencePackage ReferencePackage { get; set; }
```
Could you please use empty line between properties decalration? The same one in interfaces. I really appreciate if you could follow suggested guideline.

HenadziBaranau commented on NugetAnalyzer.Domain/Project.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public class Project
+    {
+        public int ProjectId { get; set; }
```
I prefer to use just `Id`. You are duplicating semantic prefix. Imagine code usage:
```
var project = new Project();
project.ProjectId = 1; 
// project.Id = 1; I would say it's more elegant.
```

maximKalinov replied:
Done

HenadziBaranau commented on NugetAnalyzer.Domain/PackageBase.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace NugetAnalyzer.Domain
+{
+    public abstract class PackageBase
```
Try to avoid inheritance in EF Core. EF Core currently only supports [TPH ](https://www.learnentityframeworkcore.com/inheritance) strategy, that has 2 main disadavantages:
- Data integrity(I believe that you are aware about database normalization)
- Adds additional column to the table to differ types.

Think about using(not required, but probably make sense) - [Owned Types](https://docs.microsoft.com/en-us/ef/core/modeling/owned-entities).

gennadybaranov replied:
That's fine. Agree with current approach. 

The only thing, please think about approach how are we going to use Version's components. 

- Do we leave them as they are and compare as regular numbers. 
- Do we want to move them to own type - Version?
- Do we create computed get property of .NET Version type?

HenadziBaranau commented on NugetAnalyzer.DAL/EntityConfigurations/PackageConfiguration.cs:
```diff
@@ -0,0 +1,22 @@
+                .HasOne(p => p.ReferencePackage)
+                .WithMany(p => p.CurrentPackages)
+                .HasForeignKey(p => p.ReferencePackageId)
+                .OnDelete(DeleteBehavior.SetNull);
```
Is it required? Why not Restrict or Cascade? Use cases when it's applicable?

maximKalinov replied:
Changed it on DeleteBehavior.Restrict

gennadybaranov replied:
@maximKalinov why DeleteBehavior.Restrict? :)

HenadziBaranau reviewed (changes requested):
BTW, could we remove `SetCompatibilityVersion` in Stratup? Also please allign Startup builder similar to entity type configurations

maximKalinov committed:
T2: Refactoring

maximKalinov committed:
T2: Change type of property PublishedDate to nullable

maximKalinov committed:
T2: Add connection string

maximKalinov committed:
T2: Remove unuse dbsets

maximKalinov committed:
T2: Change ConnectionString

maximKalinov committed:
T2: Add init migration

gennadybaranov commented on NugetAnalyzer/appsettings.json:
```diff
@@ -4,5 +4,8 @@
     }
   },
+  "ConnectionString": {
+    "DefaultConnection": "Data Source=Test;Initial Catalog=Dev5;Integrated Security=True;"
```
It's better to use localhost for development purporses. Your approach forces all developers create SQL instance with name Test that is not suitable.
Also please find better name for database :)

maximKalinov replied:
Done

gennadybaranov commented on NugetAnalyzer/Startup.cs:
```diff
@@ -12,16 +13,21 @@ public class Startup
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+            services.AddDbContext<NugetAnalyzerDbContext>(options =>
```
There is new feature - DbContext pooling. Please use [it](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-2.0#dbcontext-pooling).

maximKalinov replied:
Done

maximKalinov committed:
T2: Change connection string

maximKalinov committed:
T2: Change registering a custom DbContext type
